crates/wasmparser/src/validator/core.rs

```diff
@@ -973,34 +973,26 @@ mod arc {
     use std::sync::Arc;
 
     pub struct MaybeOwned<T> {
-        owned: bool,
         arc: Arc<T>,
     }
 
     impl<T> MaybeOwned<T> {
-        #[allow(clippy::cast_ref_to_mut)]
         fn as_mut(&mut self) -> Option<&mut T> {
-            if !self.owned {
-                return None;
-            }
-            debug_assert!(Arc::get_mut(&mut self.arc).is_some());
-            Some(unsafe { &mut *(Arc::as_ptr(&self.arc) as *mut T) })
+            Arc::get_mut(&mut self.arc)
```

Comment:
I mentioned this on Zulip as well, but I think it's worth measuring the performance impact of this since the original purpose of this was to avoid atomics in release mode for `assert_mut` here. If the performance impact is negligible though it seems reasonable to get rid of the `unsafe` block.